src/main/java/io/cryostat/diagnostic/BucketedDiagnosticsMetadataService.java

```diff
@@ -0,0 +1,146 @@
+import io.quarkus.runtime.StartupEvent;
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.enterprise.event.Observes;
+import jakarta.inject.Inject;
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+import org.jboss.logging.Logger;
+import software.amazon.awssdk.core.sync.RequestBody;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.DeleteObjectRequest;
+import software.amazon.awssdk.services.s3.model.GetObjectRequest;
+import software.amazon.awssdk.services.s3.model.ListObjectsV2Request;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+
+@ApplicationScoped
+@LookupIfProperty(
+        name = ConfigProperties.STORAGE_METADATA_THREAD_DUMPS_STORAGE_MODE,
+        stringValue = DiagnosticsHelper.METADATA_STORAGE_MODE_BUCKET)
+public class BucketedDiagnosticsMetadataService
+        implements CRUDService<String, ThreadDump, ThreadDump> {
+
+    @ConfigProperty(name = ConfigProperties.STORAGE_METADATA_THREAD_DUMPS_STORAGE_MODE)
+    String storageMode;
+
+    @ConfigProperty(name = ConfigProperties.AWS_BUCKET_NAME_METADATA)
+    String bucket;
+
+    @ConfigProperty(name = ConfigProperties.AWS_METADATA_PREFIX_THREAD_DUMPS)
+    String prefix;
+
+    @Inject S3Client storage;
+    @Inject StorageBuckets buckets;
+
+    @Inject ObjectMapper mapper;
+
+    @Inject Logger logger;
+
+    void onStart(@Observes StartupEvent evt) {
+        if (!DiagnosticsHelper.METADATA_STORAGE_MODE_BUCKET.equals(storageMode)) {
+            return;
+        }
+        buckets.createIfNecessary(bucket);
+    }
+
+    @Override
+    public List<ThreadDump> list() throws IOException {
+        var builder = ListObjectsV2Request.builder().bucket(bucket).prefix(prefix);
+        var objs = storage.listObjectsV2(builder.build()).contents();
+        return objs.stream()
+                .map(
+                        t -> {
+                            // TODO this entails a remote file read over the network and then some
+                            // minor processing of the received file. More time will be spent
+                            // retrieving the data than processing it, so this should be
+                            // parallelized.
+                            try {
+                                return read(t.key()).orElseThrow();
+                            } catch (IOException e) {
+                                logger.error(e);
+                                return null;
+                            }
+                        })
+                .filter(Objects::nonNull)
+                .toList();
+    }
+
+    @Override
+    public void create(String k, ThreadDump threadDump) throws IOException {
+        storage.putObject(
+                PutObjectRequest.builder()
+                        .bucket(bucket)
+                        .key(prefix(k))
+                        .contentType(HttpMimeType.JFC.mime())
```

Comment:
This seems wrong - maybe this should just be `text/plain` or something like that. `.jfc` is the XML schema for JFR event configurations (event templates).